### PR TITLE
Replace grouped-parameter lists with new syntax for argument groups.

### PR DESCRIPTION
**Pull request by toddjonker**
The new syntax `(: ...)` denotes an "argument group" and can only occur in an E-expression.  It replaces the use of lists to pass multiple expressions to a single parameter.

"Grouped parameters" are removed from signatures, leaving only cardinality. To use multiple expressions with a parameter, the caller uses an argument group.  A single expression (of appropriate shape) is always acceptable, regardless of cardinality.

In the template language, macro invocations can use the syntax `(; ...)` to denote argument groups.

As you can see from the text changes, this is more simple than the prior syntax and semantics.  It turns grouping into a syntactic concern at the macro invocation site.

### Issue #, if available:

This change is related to #291 but has at least one key difference. As proposed here, argument groups are not general-purpose streams, and can only appear in macro-argument positions.  This is because they have a critical difference from the `values` macro: their contents are shape-constrained by the relevant macro parameter's type, most obviously for macro-shaped parameters.  Personally, I'm in favor of keeping argument groups and `values` invocations syntactically distinct, to reduce confusion.

### Other notes:

* Content explaining when groups are allowed needs to in sync with the binary encoding: text E-expressions should only allow groups where the binary form allows them.
* In editing this I discovered that these docs currently don't have description of text E-expressions; I'll have to circle back and get that content added.

----
_**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**_

### CODE REVIEW

File: src/glossary.adoc
Comment (popematt) [inline src/glossary.adoc:19]: Suppose I have macros named "foo" and "bar". How do we prevent this ambiguity?
```
(:foo (:bar baz))
```
Is this an invocation of `foo` with an expression group of `'bar' 'baz'` or is it an invocation of `foo` where the argument is the result from the invocation of `(:bar baz)`?

File: src/macros-by-example.adoc
Comment (popematt) [inline src/macros-by-example.adoc]: Nit
```suggestion
there is a special delimiting syntax for an _argument group_ collecting the
```

File: src/macros-by-example.adoc
Comment (popematt) [inline src/macros-by-example.adoc:757]: What are the implications of "that align with the parameter type"? Does this require type checking of expression groups in the macro evaluator? (Seems like this might have some of the trouble of type vs encoding.)

File: src/macros-by-example.adoc
Comment (popematt) [inline src/macros-by-example.adoc:780]: Why not? We already have rules that define how to combine a stream in an expression group, allowing things like `(:prices (: 1 (:values 2 3) 4) GBP)`.

Is it legal to nest expression groups for non-"rest" parameters? E.g.: `(:prices (: 1 (: 2 3) 4) GBP)`

File: src/macros-by-example.adoc
Comment (popematt) [inline src/macros-by-example.adoc]: 1. What would you think about having some syntax in the text encoding that is specific for macro-type arguments... something that looks like a macro invocation but has some symbol that indicates that the name is redundant in that position? (Strawman suggestion: `(:*` ... `)`.) It might help improve readability in cases where there are also s-exps in the macro arguments.
2. Suppose `point` can accept multiple coordinates, then one could have something like this?
```
(:scatterplot (: (3 (:17 18 19)) (395 23) (15 48) (2023 5)) "hour" "widgets")
```

File: src/glossary.adoc
Comment (zslayton) [reply src/glossary.adoc:19]: I talked this over with Matt offline. The solution--that whitespace following `:` is the differentiator--feels like a big stumbling block. I think newcomers would often incorrectly consider:
```
(:foo (:bar baz))
```
and
```
(:foo (: bar baz))
```
to be equivalent. To be clear, I think this is just a syntactic issue. I don't object to the underlying concept of expression groups being their own construct.

I wonder if a different `[`-based enclosing syntax would work for this--say, robot smiley (`[:`), Kleene star (`[*`), or an angle bracket (`[>`).

|Syntax | E-exp arg | Expr group arg |
|-------------------|---|---|
| `(: ... )` as expr group |  `(:foo (:bar baz))` |  `(:foo (: bar baz))`  |
|`[: ... ]` as expr group |  `(:foo (:bar baz))` |  `(:foo [:bar baz])` |
|`[* ... ]` as expr group |  `(:foo (:bar baz))` |  `(:foo [*bar baz])` |
|`[> ... ]` as expr group |  `(:foo (:bar baz))` |  `(:foo [>bar baz])` |

These are more visually distinct from macro invocations IMO.

We could also include the sigil on the closing `]` if that's helpful.

|Syntax | E-exp arg | Expr group arg |
|-------------------|---|---|
|`[: ... :]` as expr group |  `(:foo (:bar baz))` |  `(:foo [:bar baz:])` |
|`[* ... *]` as expr group |  `(:foo (:bar baz))` |  `(:foo [*bar baz*])` |
|`[> ... <]` as expr group |  `(:foo (:bar baz))` |  `(:foo [>bar baz<])` |

File: src/macros-by-example.adoc
Comment (zslayton) [inline src/macros-by-example.adoc]: Refreshingly simple. :)

File: src/macros-by-example.adoc
Comment (zslayton) [reply src/macros-by-example.adoc:780]: I think your question is conflating expression groups and `(:values ...)`, which this PR hopes to make into distinct entities.

> Is it legal to nest expression groups for non-"rest" parameters? E.g.: `(:prices (: 1 (: 2 3) 4) GBP)`

No, this isn't legal; expression groups can only appear in argument positions. It's easier to understand why when you consider the binary encoding.  Say this is the definition of your macro (borrowed from the doc):

```
(macro prices
  [(amount number*),(currency symbol)]
  // ...BODY EXPR...
)
```

When the binary reader encounters the `amount` argument, it will know to decode an expression group and decode its contents: a sequence of tagged values that must be `number`s. There is no opcode for "expression group", so there's no way to encode `(: 2 3)` within the outer expression group.

I think we _could_ allow an explicit expression group for a rest parameter in text, but I don't think we want to.

>> Rest parameters effectively use implicit groups at the call site, and they cannot be
invoked with explicit groups.

> Why not? We already have rules that define how to combine a stream in an expression group, allowing things like (:prices (: 1 (:values 2 3) 4) GBP).

We could permit it semantically, but it would be strictly redundant syntactically. As an example, let's flip `currency` and `amount` so the latter can be a rest parameter.

```
(macro prices
  [(currency symbol) (amount ...number)]
  // ...BODY EXPR...
)

(:prices USD 1 2 3 4)
(:prices USD (: 1 2 3 4)) // if we allowed "explicit rest"
```
We'd need to add an extra branch to support this, and I don't think it buys us much. In fact, I think that someone passing an explicit expression group to a `...` parameter is a helpful indication that they've misunderstood something and an opportunity to emit a helpful error message.

File: src/macros-by-example.adoc
Comment (zslayton) [reply src/macros-by-example.adoc]: I think this is something else that a more visually distinctive expression group would help to address.

> What would you think about having some syntax in the text encoding that is specific for macro-type arguments... something that looks like a macro invocation but has some symbol that indicates that the name is redundant in that position?

Something I really like about the expression group is that its contents are always--for any encoding, including macro-shaped--tied to the corresponding parameter's declared encoding. When an exactly-one parameter has a macro-shaped encoding, the text representation of that encoding is an s-expression containing the arguments to the declared macro. 

_(These examples use signature syntax from recent discussions instead of the as-specced syntax)_
```
(macro point2d (x y) /* body expr */)
(macro foo (point2d::x) /* body expr */)

(:foo (1 2))
```

When there's a macro-shaped expression group, it's an expression group that contains zero or more instances of the same sexp-of-args syntax.

```
(macro point2d (x y) /* body expr */)
(macro foo (point2d::x*) /* body expr */)

(:foo (: (1 2) (3 4) (5 6) ))
```

I think making macro-shaped expression groups into yet another construct would be added complexity for limited benefit.

File: src/glossary.adoc
Comment (toddjonker) [reply src/glossary.adoc:19]: My intention was that whitespace or `)` must follow. I'm also inclined to spec, at least early on, that a macro-ref can _only_ be followed by whitespace or `)`.  This is intended to avoid extending the tokenization weirdness we already have inside S-expressions and make the E-expression "opening" visually separate from its content.

I don't think there's ambiguity; it's in the spec that the characters immediately after `(:` must be a macro reference or `)`, else there's an error; this adds a distinct case for whitespace.  A user who mistakes the difference would likely learn from the immediate syntax error.

I'm not sure why the `[` markers would be more intuitive.  In the examples here I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization, eg `[:bar@foo]`.

Adding a closing marker introduces another question regarding empty-group;  `[::]` feels off to me.

_Perhaps_ the way to slice this is to change the requirement that the leading colon must abut the paren, and instead tokenize it with the following identifier/integer.  Thus allowing `( :macro ...)`.  Then we can teach that the colon is either (and only) the start of the macro-ref or a group.  But I hesitate because (a) elsewhere in Ion "whitespace" includes commas, and (b) this means the parser may need to scan an unbounded amount of "whitespace" looking for that colon.  This is already a problem with annotations, and I'm loathe to repeat it.

File: src/macros-by-example.adoc
Comment (toddjonker) [reply src/macros-by-example.adoc]: I think adding syntax for macro-encoded arguments defeats the purpose, which is to improve brevity.  I think

```
(:foo (: (:* 1 2) (:* 3 4) (:* 5 6)))
```
is significantly less human-friendly than:
```
(:foo (: (1 2) (3 4) (5 6)))
```

The concept also
* intersects with the ungrouped case: does one write `(:foo (1 2))` or `(:foo (:* 1 2))`?
* requires additional tokenization rules following the `(:*`

File: src/macros-by-example.adoc
Comment (toddjonker) [reply src/macros-by-example.adoc:780]: @zslayton expresses my thinking well.  In particular, this is great:

> There is no opcode for "expression group"

File: src/macros-by-example.adoc
Comment (toddjonker) [reply src/macros-by-example.adoc:757]: It should probably read "base encoding" or something; the terminology is muddled and inconsistent.  Also this would be clarified in the missing section tracked by #303 ... we somehow missed the section defining text E-expression syntax that would be affected by this PR.

File: src/glossary.adoc
Comment (zslayton) [reply src/glossary.adoc:19]: > I don't think there's ambiguity; it's in the spec that the characters immediately after (: must be a macro reference or ), else there's an error; this adds a distinct case for whitespace. A user who mistakes the difference would likely learn from the immediate syntax error.

It's not ambiguous for a parser, but it's ambiguous if you're a human. :)

> I'm not sure why the [ markers would be more intuitive. In the examples here I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization, eg [:bar@foo]

I'm not prescribing a solution, I'm illustrating that having a syntax wherein expressions as similar as `(:foo)` and `(: foo)` having two different meanings is readily avoidable. Users will definitely learn of their mistake when they stub their toe on it, but I think we could spare them that trouble.

> Perhaps the way to slice this is to change the requirement that the leading colon must abut the paren,

I would also prefer not to because of the reasons you stated.

Comment (popematt) [reply src/glossary.adoc:19]: > I'm not sure why the `[` markers would be more intuitive.

I'm not looking for "more intuitive"—just something that is less "potential foot-gun" than `(: foo)` vs `(:foo)`.

It just so happens that `[:` is currently not legal Ion, so it's something that _could_ be used.

What I find incredibly *un*intuitive is that these are all equivalent:
* `(foo)` and `( foo)`
* `[foo]` and `[ foo]`
* `{foo:1}` and `{ foo:1}`
* `{{abc12=}}` and `{{ abc12=}}`
 
...but `(:foo)` and `(: foo)` are not equivalent _and_ they are both valid Ion.

> I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization

If we do use `[:` or similar, then I think commas are fine.

If we're concerned about tokenization simplicity or efficiency, then why not use something like `[:`. Instead of having to check three characters to determine whether we're looking at the start of a macro vs an expression group, we only have to look at two characters—`(: ` vs `[:`. If we know because of the context that it _must_ be either a macro or expression group, then we only actually need to look at the first character—`(` or `[`—to determine which it is.

Comment (popematt) [reply src/glossary.adoc:19]: Sorry if I'm belaboring this point, but I thought of one last thing. This will be the last one, I promise.

> Users will definitely learn of their mistake when they stub their toe on it, but I think we could spare them that trouble.

Ideally, we have lots of Ion users (who could each potentially make this mistake), and even if users learn and are aware of it, it's easy to accidentally add a single whitespace. We should assume that users (collectively) will make this mistake over and over, so we should try to minimize the potential impact of and time to identify the mistake. 

When a user accidentally changes e.g. `(:foo "bar-" (:baz a b c))` to `(:foo "bar-" (: baz a b c))`:
* If argument groups start with `[:` (or similar), then they _only_ stub their toe because the result is invalid Ion. A syntax highlighter can tell you immediately that it's syntactically wrong and exactly where it is wrong.
* If argument groups start with `(: `, then the user potentially shoots their foot off by producing new data that is valid Ion but different than what they intended. It _might_ end up failing a cardinality check somewhere, but that's not guaranteed. Valid, subtly different data can only reliably be detected by a tool (or person) that is capable of understanding the application-defined semantics of the data. The problem might not be detected right away, and once it is identified that there is a problem, it won't necessarily be obvious where the mistake was made (especially if the "new" data is passed around between multiple systems or transformed in some way before a discrepancy is noticed).

If you use one of `[:`, `[>`, etc. _and_ have commas between elements in the argument group, then it's even more obvious when you have a macro vs an argument group, and it's significantly more difficult to turn an argument group into a macro invocation (or vice versa) by accident.

Comment (popematt) [reply src/glossary.adoc:19]: I think it would help me understand your point of view if you can explain why you think `(: ` is the best syntax for argument groups.

Comment (toddjonker) [reply src/glossary.adoc:19]: There's been a lot of good discussion on the syntax, but I'd like to avoid blocking this PR on it.  Changing the concrete syntax with low impact (since there's currently no usage) and happen next, but many other low-level things pivot on moving the grouping concept from parameter-declaration time to invocation-time.  I'd like to unblock that work by getting this merged while leaving the syntax open for discussion.

**Review by popematt — APPROVED** [review]

**Comment by toddjonker** [comment]
I believe we have consensus, modulo the concrete syntax.  Holding off on merging to let people review my updates yesterday.

**Comment by toddjonker** [comment]
Ooops @popematt our messages passed in spacetime.  Did you get a chance to see the latest updates?

**Comment by popematt** [comment]
@toddjonker Yes, I did see the updates.

File: src/glossary.adoc
Comment (toddjonker) [reply src/glossary.adoc:19]: It occurs to me that another potential syntax for argument groups is `(, ...)` since comma, like colon, cannot appear unquoted in an S-expression.  It arguably carries the correct connotation of "sequence of elements", and might be better aligned with the current template-language equivalent `(; ...)`

I would not allow comma inside such a comma-group. I would want everything but the comma to follow S-expression tokenization rules, so that it's consistent with the TL syntax.

**Pull request merged**
